src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
+}
+
+
+static void init()
+{
+    static std::once_flag init_flag;
+    std::call_once(init_flag, init_callback);
+}
+
+static std::string sha256(size_t data_size, const void* data)
+{
+    BCRYPT_ALG_HANDLE hAlg = NULL;
+    BCRYPT_HASH_HANDLE hHash = NULL;
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    std::string errorString;
+    DWORD cbData = 0;
+    DWORD cbHash = 0;
+    DWORD cbHashObject = 0;
+    PBYTE pbHashObject = NULL;
+    PBYTE pbHash = NULL;
+    std::string retval;
+
+    //open an algorithm handle
+    if (!NT_SUCCESS(status = BCryptOpenAlgorithmProvider(
+        &hAlg,
+        BCRYPT_SHA256_ALGORITHM,
+        NULL,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptOpenAlgorithmProvider\n";
+        goto Cleanup;
+    }
+
+    if (!NT_SUCCESS(status == BCryptGetProperty(
+        hAlg,
+        BCRYPT_OBJECT_LENGTH,
+        (PBYTE)&cbHashObject,
+        sizeof(DWORD),
+        &cbData,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptGetProperty\n";
+        goto Cleanup;
+    }
+
+    pbHashObject = (PBYTE)HeapAlloc(GetProcessHeap(), 0, cbHashObject);
+    if (pbHashObject == NULL)
+    {
+        errorString = "Memory allocation failed\n";
+        status = STATUS_NO_MEMORY;
+        goto Cleanup;
+    }
+
+    //calculate the length of the hash
+    if (!NT_SUCCESS(status = BCryptGetProperty(
+        hAlg,
+        BCRYPT_HASH_LENGTH,
+        (PBYTE)&cbHash,
+        sizeof(DWORD),
+        &cbData,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptGetProperty\n";
+        goto Cleanup;
+    }
+
+    //allocate the hash buffer on the heap
+    pbHash = (PBYTE)HeapAlloc(GetProcessHeap(), 0, cbHash);
+    if (NULL == pbHash)
+    {
+        errorString = "Memory allocation failed\n";
+        status = STATUS_NO_MEMORY;
+        goto Cleanup;
+    }
+
+    //create a hash
+    if (!NT_SUCCESS(status = BCryptCreateHash(
+        hAlg,
+        &hHash,
+        pbHashObject,
+        cbHashObject,
+        NULL,
+        0,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptCreateHash\n";
+        goto Cleanup;
+    }
+
+
+    //hash some data
+    if (!NT_SUCCESS(status = BCryptHashData(
+        hHash,
+        (PBYTE)data,
+        (ULONG)data_size,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptHashData\n";
+        goto Cleanup;
+    }
+
+    //close the hash
+    if (!NT_SUCCESS(status = BCryptFinishHash(
+        hHash,
+        pbHash,
+        cbHash,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptFinishHash\n";
+        goto Cleanup;
+    }
+
+    retval.reserve(2 * cbHash + 1);
+    for (size_t i = 0; i < cbHash; i++)
+    {
+        char buf[3];
+        snprintf(buf, sizeof(buf), "%02x", pbHash[i]);
```

Comment:
Why use sprintf here? Why not manually create the string?

Reply:
How it is done in the Linux version